analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs

```diff
@@ -295,9 +297,15 @@ private static ProgramState InitLocals(ControlFlowBranch branch, ProgramState st
         branch.IsConditionalSuccessor ? condition == constraint : condition != constraint;
 
     private static bool IsReachable(ControlFlowRegion region, ExceptionState thrown) =>
-        region.Kind == ControlFlowRegionKind.Finally
+        CatchesAll(region)
         || thrown == ExceptionState.UnknownException
-        || region.ExceptionType.Is(KnownType.System_Object)       // catch when (condition)
-        || region.ExceptionType.Is(KnownType.System_Exception)
         || thrown.Type.DerivesFrom(region.ExceptionType);
+
+    private static bool CatchesAll(ControlFlowRegion region) =>
```

Comment:
The fact that `Finally` answers `true` to `CatchesAll` is a bit weird. What about not dealing with finally here, but on the invocation side (with duplication) to make it more clear?